src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -63229,6 +63229,192 @@ void M0(U x0)
+                // (7,9): hidden CS8607: Expression is probably never null.
+                //         x0 ??= y0;
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "x0").WithLocation(7, 9),
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         M1(x0).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "M1(x0)").WithLocation(8, 9),
+                // (10,9): hidden CS8607: Expression is probably never null.
+                //         x0?.ToString();
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "x0").WithLocation(10, 9));
+        }
+
+        [Fact]
+        public void NullabilityOfTypeParameters_259()
+        {
+            var source = @"
+class Outer
+{
+    void M0<T>(T x0, T y0, T a0)
+    {
+        if (x0 == null) return;
+        if (y0 == null) return;
+        x0 ??= y0;
+        M1(x0) = a0;
+        x0?.ToString();
+        M1(x0).ToString();
+    }
+
+    ref S M1<S>(S x) => throw null;
+}
+";
+
+            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
+                // (8,9): hidden CS8607: Expression is probably never null.
+                //         x0 ??= y0;
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "x0").WithLocation(8, 9),
+                // (10,9): hidden CS8607: Expression is probably never null.
+                //         x0?.ToString();
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "x0").WithLocation(10, 9),
+                // (11,9): warning CS8602: Possible dereference of a null reference.
+                //         M1(x0).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "M1(x0)").WithLocation(11, 9));
+        }
+
+        [Fact]
+        public void NullabilityOfTypeParameters_260()
+        {
+            var source = @"
+class Outer<T>
+{
+    void M0(T x0, T y0)
+    {
+        if (y0 == null) return;
+        y0 ??= M2();
+        M1(y0) = x0;
+        M1<T>(y0) = x0;
+
+        M1(y0).ToString();
+        y0.ToString();
+    }
+
+    ref S M1<S>(S x) => throw null;
+
+#nullable disable
+    T M2() => throw null;
+}
+";
+
+            CreateCompilation(source, options: WithNonNullTypesTrue()).VerifyDiagnostics(
+                // (7,9): hidden CS8607: Expression is probably never null.
+                //         y0 ??= M2();
+                Diagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, "y0").WithLocation(7, 9),
+                // (11,9): warning CS8602: Possible dereference of a null reference.
+                //         M1(y0).ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "M1(y0)").WithLocation(11, 9));
+        }
+
+        [Fact]
+        public void NullabilityOfTypeParameters_261()
+        {
+            var source = @"
+class Outer<T>
+{
+    void M0(T x0, T y0)
+    {
+        y0 ??= M2();
```

Comment:
>y0 ??= M2(); [](start = 8, length = 12)

It would also be good to add test around unconstrained type parameters to verify that we properly warn on issues around the assignment part.